Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISSumBanks.py

```diff
@@ -0,0 +1,110 @@
+        self.declareProperty(
+            MatrixWorkspaceProperty(self._WORKSPACE, '', direction=Direction.Input, optional=PropertyMode.Mandatory),
+            doc='An input 2D detector workspace')
+        self.declareProperty(self._ROI, defaultValue="",
+                             doc="List of detector IDs to include")
+        self.declareProperty(
+            MatrixWorkspaceProperty(self._OUTPUT_WS, '', direction=Direction.Output),
+            doc='The preprocessed output workspace. If multiple input runs are specified '
+                'they will be summed into a single output workspace.')
+
+    def PyExec(self):
+        input_workspace = self.getProperty(self._WORKSPACE).value
+        roi_detector_ids = self.getProperty(self._ROI).value
+
+        masked_workspace = self.mask_detectors(input_workspace, roi_detector_ids)
+        summed_workspace = self.sum_banks(masked_workspace)
+
+        self.setProperty(self._OUTPUT_WS, summed_workspace)
+
+    def validateInputs(self):
+        issues = dict()
+        workspace = self.getProperty(self._WORKSPACE).value
+        try:
+            self._get_rectangular_detector_component(workspace)
+        except Exception as err:
+            issues['InputWorkspace'] = str(err)
+        return issues
+
+    def mask_detectors(self, workspace: MatrixWorkspace, roi_detector_ids: str) -> MatrixWorkspace:
+        # We need to apply the mask to the original workspace so we can extract a MaskWorkspace to use in
+        # BinaryOperateMasks which inverts it. First, we take a clone to not destructively alter the original WS
+        cloned_ws = self._run_child_with_out_props("CloneWorkspace", InputWorkspace=workspace)
+        self.createChildAlgorithm('MaskDetectors', Workspace=cloned_ws, DetectorList=roi_detector_ids).execute()
+        mask_ws = self._run_child_with_out_props('ExtractMask', InputWorkspace=cloned_ws)
+
+        self.createChildAlgorithm('BinaryOperateMasks', InputWorkspace1=mask_ws,
+                                  OperationType='NOT', OutputWorkspace=mask_ws).execute()
+
+        # and then re-apply this to the original workspace, again taking a clone
+        cloned_ws = self._run_child_with_out_props("CloneWorkspace", InputWorkspace=workspace)
+        self.createChildAlgorithm('MaskDetectors', Workspace=cloned_ws, MaskedWorkspace=mask_ws).execute()
+        return cloned_ws
+
+    def sum_banks(self, workspace: MatrixWorkspace):
+        component = self._get_rectangular_detector_component(workspace)
+        num_banks = component.xpixels()
+        return self._run_child_with_out_props('SmoothNeighbours', InputWorkspace=workspace,
+                                              SumPixelsX=num_banks, SumPixelsY=1)
+
+    def _get_rectangular_detector_component(self, workspace: MatrixWorkspace):
+        result = None
+        inst = workspace.getInstrument()
+        if not inst:
+            raise RuntimeError('The input workspace must have an instrument')
```

Comment:
This should be checked in validate properties since it's a fairly cheap operation (just a get) to do twice

Reply:
This function will be called by validateInputs so it should be caught there...

Reply:
:man_facepalming: Good point :+1: